File: frontend/src/lib/stores/projects.ts
/** Global project state — the operator's working directories and which one is
 *  active. One place owns it so the nav switcher, the projects screens, and the
 *  chat composer's code-mode picker share a single source of truth.
 *
 *  **The selection is the backend's, not ours.** `POST /projects/{id}/activate`
 *  persists it, and the response carries the whole listing back, so activating
 *  reseats from one shape rather than patching a local guess. What lives here is
 *  the echo plus the header the client sends (`~/lib/api/projectScope`), which only
 *  ever relays that selection — the backend decides what a scope means and
 *  re-resolves it on every request.
 *
 *  **Null is a real state, not an empty one.** No active project means the operator
 *  sees their unfiled work: exactly what the app showed before projects existed.
 *  Activating one *adds* that project's items to the unfiled ones; it never hides
 *  them. Another project's items are what stay invisible. */

import {
  createEffect,
  createResource,
  createRoot,
  createSignal,
  type Resource,
} from "solid-js";
import { api } from "~/lib/api";
import { ALL_PROJECTS, setProjectScope } from "~/lib/api/projectScope";
import { useSession } from "~/lib/stores/session";

/** The git facts, re-probed by the backend on every listing. */
export interface ProjectRepo {
  exists: boolean;
  isGitRepo: boolean;
  /** Null when the path is not a repo — distinct from 0, which means a clean tree. */
  uncommittedChanges: number | null;
  currentBranch: string | null;
}

export interface Project {
  id: string;
  name: string;
  rootPath: string;
  gitInitialized: boolean;
  baseRef: string;
  archived: boolean;
  createdAt: string;
  lastOpenedAt: string;
  repo: ProjectRepo;
}

interface ProjectsDTO {
  projects: Project[];
  activeId: string | null;
}

async function fetchProjects(): Promise<ProjectsDTO> {
  return api.get<ProjectsDTO>("/projects");
}

/** Activation bookkeeping — deliberately plain variables rather than signals, so
 *  reading them can never make a computation re-run.
 *
 *  `POST /projects/{id}/activate` is a round trip, and the operator picks again as
 *  soon as the switcher looks like it ignored them. Both replies then arrive, and
 *  whichever lands last writes the echo *and* the `X-Ody-Project` header every
 *  subsequent request is scoped by — so a slow first pick can silently re-scope the
 *  whole app to the project they abandoned, taking every refetch with it. Only the
 *  operator's newest pick is allowed to adopt its own answer; an older reply is read
 *  and dropped. `activations` is the same fact for the listing effect above: how many
 *  writes are still in the air. */
let picks = 0;
let activations = 0;

const store = createRoot(() => {
  const session = useSession();

  const [tick, setTick] = createSignal(1);
  const [data] = createResource(
    () => (session.isAuthenticated ? tick() : false),
    fetchProjects,
  );

  // The active id, held as a local echo so the switcher moves at click speed. Seeded
  // and reconciled from the backend listing, which is authoritative.
  const [activeId, setActiveId] = createSignal<string | null>(null);
  // True only while the operator has explicitly asked for ALL PROJECTS. Distinct from
  // "nothing is active": the backend reads an absent header as "use the stored
  // selection", so saying *unscoped* takes a literal.
  const [unscoped, setUnscoped] = createSignal(false);

  // Seed the echo from every listing the backend returns — not just from the writes
  // that happen to return one. Without this the echo stays null across a reload while
  // the backend is still scoped to its stored selection, so the switcher reads
  // "UNFILED ONLY" and no project row looks selected while the operator is in fact
  // inside a project. Worse, toggling ALL PROJECTS and back would then write that null
  // into the request header and genuinely change the scope to match the wrong label.
  createEffect(() => {
    const dto = data.latest;
    if (!dto || unscoped()) return;
    // Not while an activation is in the air. A listing fetched before the write
    // describes the selection the operator is in the middle of leaving, and adopting
    // it here would put the old project back into the scope header for every request
    // that follows. The activation reseats from its own response and ticks a fresh
    // listing, so nothing is lost by skipping this one. Read untracked on purpose:
    // this effect must not re-run when the count falls, or it would adopt the very
    // listing it just declined.
    if (activations > 0) return;
    setActiveId(dto.activeId);
    setProjectScope(dto.activeId);
  });

  return { data, tick, setTick, activeId, setActiveId, unscoped, setUnscoped };
});

/** The listing, including archived-excluded projects. */
export function useProjects(): Resource<ProjectsDTO> {
  return store.data;
}

export function activeProjectId(): string | null {
  return store.unscoped() ? null : store.activeId();
}

export function isUnscoped(): boolean {
  return store.unscoped();
}

/** The active project's full record, or null when nothing is active. */
export function activeProject(): Project | null {
  const id = activeProjectId();
  if (!id) return null;
  return store.data.latest?.projects.find((p) => p.id === id) ?? null;
}

export function refreshProjects(): void {
  store.setTick((n) => n + 1);
}

/** Reconcile the local echo from a backend listing. Called after any write that
 *  returns one, so the two can't drift. */
function adopt(dto: ProjectsDTO): void {
  store.setActiveId(dto.activeId);
  setProjectScope(store.unscoped() ? ALL_PROJECTS : dto.activeId);
}

/** Switch the active project, or pass `null` to clear it. The backend persists the
 *  selection and returns the whole listing; we reseat from that — unless the operator
 *  has since picked something else, in which case this answer is stale the moment it
 *  arrives and adopting it would undo their newer choice. Resolves either way: the
 *  call did reach the backend, and the surface that made it has nothing to report. */
export async function setActiveProject(
  projectId: string | null,
): Promise<void> {
  store.setUnscoped(false);
  const mine = ++picks;
  activations += 1;
  try {
    const dto = projectId
      ? await api.post<ProjectsDTO>(`/projects/${projectId}/activate`, {})
      : await api.post<ProjectsDTO>("/projects/deactivate", {});
    if (mine !== picks) return;
    adopt(dto);
    store.setTick((n) => n + 1);
  } catch (err) {
    // A superseded pick's failure is not news either: the operator has moved on, and
    // "could not switch project" over the project they *are* now in reads as a broken
    // switcher rather than as the history it is.
    if (mine === picks) throw err;
  } finally {
    activations -= 1;
  }
}

/** ALL PROJECTS — deliberately unscoped, which is a different request from having
 *  no active project. Held client-side because it is a way of *asking*, not a stored
 *  preference: the operator's actual selection stays put underneath it. */
export function setUnscoped(on: boolean): void {
  store.setUnscoped(on);
  setProjectScope(on ? ALL_PROJECTS : store.activeId());
}

export async function createProject(
  name: string,
  rootPath: string,
): Promise<Project> {
  const created = await api.post<Project>("/projects", { name, rootPath });
  refreshProjects();
  return created;
}

/** The project for a directory, filed on first use.
 *
 *  What lets a code session start from *any* folder instead of from a project the
 *  operator had to create first. Idempotent by the directory itself — the backend
 *  matches on the resolved path — so pointing at the same place twice returns the
 *  same project and the same worktree machinery rather than a second row cutting a
 *  second branch from one repository. */
export async function ensureProjectForPath(rootPath: string): Promise<Project> {
  const project = await api.post<Project>("/projects/ensure", { rootPath });
  refreshProjects();
  return project;
}

export async function updateProject(
  id: string,
  patch: { name?: string; baseRef?: string; archived?: boolean },
): Promise<Project> {
  const updated = await api.patch<Project>(`/projects/${id}`, patch);
  refreshProjects();
  return updated;
}

export async function deleteProject(id: string): Promise<void> {
  await api.del(`/projects/${id}`);
  // Deleting the active project clears the selection backend-side; re-read rather
  // than guessing which way it went.
  const dto = await fetchProjects();
  adopt(dto);
  refreshProjects();
}

File: backend/routes/projects.py
"""The projects surface — the operator's working directories and the active selection.

Thin over `services/projects`. Two shapes deserve a note:

`repo` on every listing is **probed live**, not stored. Whether a directory is a git
repository, and how many changes are uncommitted, are facts about the world that change
without us; caching them would mean showing the operator a stale answer to the one
question that decides whether code mode is safe to start.

`uncommittedChanges` is surfaced deliberately, and the UI is expected to show it. Code
mode branches a worktree from the project's base ref, so uncommitted work in the
operator's own tree is **invisible to the agent**. That is the price of never touching
their tree, and it should be read on the project screen rather than discovered halfway
through a session.

camelCase out, matching the `corpus`/`uploads`/`tasks` surfaces the frontend
seams were built against.
"""

from __future__ import annotations

from datetime import datetime
from pathlib import Path

from fastapi import APIRouter, HTTPException, Request

from core.exceptions import InvalidInputError, NotFoundError
from routes import deps
from routes.camel import CamelModel
from routes.deps import OPERATOR_ID
from services.projects import ProjectView, WorktreeError
from services.projects.listing import list_files, worktree_root

router = APIRouter(prefix="/projects", tags=["projects"])


class RepoOut(CamelModel):
    exists: bool
    is_git_repo: bool
    uncommitted_changes: int | None = None
    current_branch: str | None = None


class ProjectOut(CamelModel):
    id: str
    name: str
    root_path: str
    git_initialized: bool
    base_ref: str
    archived: bool
    created_at: datetime
    last_opened_at: datetime
    repo: RepoOut


class FileEntryOut(CamelModel):
    #: Relative to the listed root, forward slashes, and the string the operator's
    #: reference carries — so it resolves the same way wherever the run's files are.
    path: str
    name: str


class FilesOut(CamelModel):
    #: Which filesystem answered. `worktree` is the thread's own branch checkout;
    #: `project` is the operator's checkout, which is what a thread gets before its
    #: first turn has created a worktree.
    root: str
    entries: list[FileEntryOut]
    #: The scan hit its bound. Said out loud rather than implying the tree is this small.
    truncated: bool


class ProjectCreate(CamelModel):
    name: str = ""
    root_path: str


class ProjectEnsure(CamelModel):
    """A directory, with no name — the name is the directory's own. Naming a project is a
    thing the operator may do later, not a form standing between them and starting work."""

    root_path: str


class ProjectUpdate(CamelModel):
    name: str | None = None
    base_ref: str | None = None
    archived: bool | None = None


class ProjectsOut(CamelModel):
    projects: list[ProjectOut]
    #: The operator's current selection, or null when nothing is active — which means
    #: they see exactly the unfiled rows they saw before projects existed.
    active_id: str | None = None


def _out(view: ProjectView) -> ProjectOut:
    return ProjectOut(
        id=view.id,
        name=view.name,
        root_path=view.root_path,
        git_initialized=view.git_initialized,
        base_ref=view.base_ref,
        archived=view.archived,
        created_at=view.created_at,
        last_opened_at=view.last_opened_at,
        repo=RepoOut(
            exists=view.probe.exists,
            is_git_repo=view.probe.is_git_repo,
            uncommitted_changes=view.probe.uncommitted_changes,
            current_branch=view.probe.current_branch,
        ),
    )


@router.get("", response_model=ProjectsOut)
async def list_projects(request: Request, include_archived: bool = False) -> ProjectsOut:
    store = deps.projects(request)
    views = await store.list(OPERATOR_ID, include_archived=include_archived)
    return ProjectsOut(
        projects=[_out(v) for v in views],
        active_id=await store.active_id(OPERATOR_ID),
    )


@router.post("", response_model=ProjectOut, status_code=201)
async def create_project(request: Request, body: ProjectCreate) -> ProjectOut:
    try:
        return _out(await deps.projects(request).create(OPERATOR_ID, body.name, body.root_path))
    except InvalidInputError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.post("/ensure", response_model=ProjectOut)
async def ensure_project(request: Request, body: ProjectEnsure) -> ProjectOut:
    """The project for a directory, created on first use.

    What lets a code session start from *any* directory instead of from a project the
    operator had to file first. It is idempotent by the directory itself, so pointing at
    the same folder twice returns the same project and the same worktree machinery rather
    than a second row cutting a second branch from one repository.

    Declared **before** ``/{project_id}``: FastAPI matches in declaration order, so the
    dynamic route would otherwise swallow ``ensure`` as an id and answer 404.
    """
    try:
        return _out(await deps.projects(request).ensure_for_path(OPERATOR_ID, body.root_path))
    except InvalidInputError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("/{project_id}", response_model=ProjectOut)
async def get_project(request: Request, project_id: str) -> ProjectOut:
    try:
        return _out(await deps.projects(request).get(OPERATOR_ID, project_id))
    except NotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/{project_id}/files", response_model=FilesOut)
async def list_project_files(
    request: Request,
    project_id: str,
    conversation_id: str | None = None,
    query: str = "",
    limit: int = 200,
) -> FilesOut:
    """The files a code thread can reference — what the composer's `@` picker browses.

    **Rooted on the project, upgraded by the conversation.** `conversation_id` is
    optional because the operator types `@` in a composer that may not have started a
    thread yet; naming one asks for that thread's own worktree, and the resolver hands
    back the project root whenever there isn't one. The response says which answered, so
    the picker can label a listing that is the operator's checkout rather than the
    agent's view of it.

    It **never creates a worktree**: acquiring the project's single checkout as a side
    effect of typing a character would take it from whatever else holds it.

    Not on `/worktrees`, which is the natural-looking home: that router resolves a
    conversation to its project and 404s until a thread is saved with a binding, which is
    precisely the moment the picker is most useful.
    """
    try:
        project = await deps.projects(request).get(OPERATOR_ID, project_id)
    except NotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    root = Path(project.root_path)
    from_worktree = False
    if conversation_id:
        root, from_worktree = await worktree_root(
            deps.worktrees(request).path_for(project_id), conversation_id, root
        )
    listing = await list_files(root, query=query, limit=max(1, min(limit, 500)))
    return FilesOut(
        root="worktree" if from_worktree else "project",
        entries=[FileEntryOut(path=e.path, name=e.name) for e in listing.entries],
        truncated=listing.truncated,
    )


@router.patch("/{project_id}", response_model=ProjectOut)
async def update_project(request: Request, project_id: str, body: ProjectUpdate) -> ProjectOut:
    try:
        return _out(
            await deps.projects(request).update(
                OPERATOR_ID,
                project_id,
                name=body.name,
                base_ref=body.base_ref,
                archived=body.archived,
            )
        )
    except NotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.delete("/{project_id}", status_code=204)
async def delete_project(request: Request, project_id: str) -> None:
    try:
        await deps.projects(request).delete(OPERATOR_ID, project_id)
    except NotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/{project_id}/init-repo", response_model=ProjectOut)
async def init_repo(request: Request, project_id: str) -> ProjectOut:
    """Make the project's directory a git repository, with the operator's explicit yes.

    Code mode needs one — a worktree is cut from it — but running `git init` and
    committing someone's whole directory is a real, visible side effect, so it is never
    implicit. This route *is* the confirmation: the UI asks, the operator answers, and
    only then does anything happen. The agent has no path to it.
    """
    store = deps.projects(request)
    try:
        project = await store.get(OPERATOR_ID, project_id)
        created = await deps.worktrees(request).ensure_repo(
            Path(project.root_path), confirmed=True
        )
    except NotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except (InvalidInputError, WorktreeError) as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    if created:
        await store.update(OPERATOR_ID, project_id, git_initialized=True)
    return _out(await store.get(OPERATOR_ID, project_id))


@router.post("/{project_id}/activate", response_model=ProjectsOut)
async def activate_project(request: Request, project_id: str) -> ProjectsOut:
    """Make this the operator's active project. Returns the whole listing, because the
    selection changes what every other surface will return and the client should reseat
    from one shape rather than patching a local guess."""
    store = deps.projects(request)
    try:
        await store.activate(OPERATOR_ID, project_id)
    except NotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return await list_projects(request)


@router.post("/deactivate", response_model=ProjectsOut)
async def deactivate_project(request: Request) -> ProjectsOut:
    """Clear the selection — the ALL PROJECTS state, which is the app's original
    behavior and not an empty one."""
    await deps.projects(request).activate(OPERATOR_ID, None)
    return await list_projects(request)
